java/algorithm/src/main/java/org/apache/arrow/algorithm/dictionary/SearchDictionaryEncoder.java

```diff
@@ -0,0 +1,110 @@
+
+  /**
+   * The dictionary for encoding/decoding.
+   * It must be sorted.
+   */
+  private final D dictionary;
+
+  /**
+   * The criteria by which the dictionary is sorted.
+   */
+  private final VectorValueComparator<D> comparator;
+
+  /**
+   * A flag indicating if null should be encoded.
+   */
+  private final boolean encodeNull;
+
+  /**
+   * Constructs a dictionary encoder.
+   * @param dictionary the dictionary. It must be in sorted order.
+   * @param comparator the criteria for sorting.
+   */
+  public SearchDictionaryEncoder(D dictionary, VectorValueComparator<D> comparator) {
+    this(dictionary, comparator, false);
+  }
+
+  /**
+   * Constructs a dictionary encoder.
+   * @param dictionary the dictionary. It must be in sorted order.
+   * @param comparator the criteria for sorting.
+   * @param encodeNull a flag indicating if null should be encoded.
+   *     It determines the behaviors for processing null values in the input during encoding/decoding.
+   *     <li>
+   *       For encoding, when a null is encountered in the input,
+   *       1) If the flag is set to true, the encoder searches for the value in the dictionary, 
+   *       and outputs the index in the dictionary.
+   *       2) If the flag is set to false, the encoder simply produces a null in the output.
+   *     </li>
+   *     <li>
+   *       For decoding, when a null is encountered in the input, 
+   *       1) If the flag is set to true, the decoder should never expect a null in the input.
+   *       2) If set to false, the decoder simply produces a null in the output.
+   *     </li>
+   */
+  public SearchDictionaryEncoder(D dictionary, VectorValueComparator<D> comparator, boolean encodeNull) {
+    this.dictionary = dictionary;
+    this.comparator = comparator;
+    this.encodeNull = encodeNull;
+
+    boolean dictionaryContainsNull = dictionary.getNullCount() > 0;
+    if (encodeNull && !dictionaryContainsNull) {
+      throw new IllegalStateException("Null should be encoded, but no null in the dictionary.");
+    }
+  }
+
+  /**
+   * Encodes an input vector by binary search.
+   * So the algorithm takes O(n * log(m)) time, where n is the length of the input vector,
+   * and m is the length of the dictionary.
+   * @param input the input vector.
+   * @param output the output vector. Note that it must be in a fresh state. At least,
+   *     all its validity bits should be clear.
+   */
+  public void encode(D input, E output) {
+    for (int i = 0; i < input.getValueCount(); i++) {
+      if (!encodeNull && input.isNull(i)) {
+        // for this case, we should simply output a null in the output.
+        // by assuming the output vector is fresh, we do nothing here.
+        continue;
+      }
+
+      int index = VectorSearcher.binarySearch(dictionary, comparator, input, i);
+      if (index == -1) {
+        throw new IllegalStateException("The data element is not found in the dictionary: " + i);
```

Comment:
IllegalArgument is potentially better here.

Reply:
Sure. Revised.